src/NBitcoin/ChainedBlock.cs

```diff
@@ -449,13 +452,51 @@ public ChainedBlock GetAncestor(int height)
 
             ChainedBlock current = this;
 
-            while (true)
+            int currentHeight = this.Height;
+            while (currentHeight > height)
             {
-                if (current.Height == height)
-                    return current;
-
-                current = current.Previous;
+                int heightSkip = GetSkipHeight(currentHeight);
+                int heightSkipPrev = GetSkipHeight(currentHeight - 1);
+                if ((current.Skip != null) && ((heightSkip == height) || ((heightSkip > height) && !((heightSkipPrev < (heightSkip - 2)) && (heightSkipPrev >= height)))))
+                {
+                    // Only follow Skip if Previous.Skip isn't better than Skip.Previous.
+                    current = current.Skip;
+                    currentHeight = heightSkip;
+                }
+                else
+                {
+                    current = current.Previous;
+                    currentHeight--;
+                }
             }
+
+            return current;
+        }
+
+        /// <summary>
+        /// Compute what height to jump back to for the skip block given this height.
+        /// </summary>
+        /// <param name="height">Height to compute skip height for.</param>
```

Comment:
>but the following expression was taken from bitcoin core

Can you put a link to the code in core? (would be nice to have)

Reply:
Sure I can do that.